mempool/cat/pool.go

```diff
@@ -0,0 +1,786 @@
+// set priority values on transactions in the CheckTx response. When selecting
+// transactions to include in a block, higher-priority transactions are chosen
+// first.  When evicting transactions from the mempool for size constraints,
+// lower-priority transactions are evicted first. Transactions themselves are
+// unordered (A map is used). They can be broadcast in an order different from 
+// the order to which transactions are entered. There is no guarantee when CheckTx
+// passes that a transaction has been successfully broadcast to any of its peers.
+//
+// A TTL can be set to remove transactions after a period of time or a number
+// of heights.
+//
+// A cache of rejectedTxs can be set in the mempool config. Transactions that
+// are rejected because of `CheckTx` or other validity checks will be instantly
+// rejected if they are seen again. Committed transactions are also added to
+// this cache. This serves somewhat as replay protection but applications should 
+// implement something more comprehensive
+type TxPool struct {
+	// Immutable fields
+	logger       log.Logger
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+	metrics      *mempool.Metrics
+
+	updateMtx            sync.Mutex
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // one value sent per height when mempool is not empty
+	preCheckFn           mempool.PreCheckFunc
+	postCheckFn          mempool.PostCheckFunc
+	height               int64 // the latest height passed to Update
+
+	// Thread-safe cache of rejected transactions for quick look-up
+	rejectedTxCache *LRUTxCache
+	// Thread-safe cache of valid txs that were evicted
+	evictedTxs *EvictedTxCache
+	// Thread-safe list of transactions peers have seen that we have not yet seen
+	seenByPeersSet *SeenTxSet
+
+	// Store of wrapped transactions
+	store *store
+
+	// broadcastCh is an unbuffered channel of new transactions that need to
+	// be broadcasted to peers. Only populated if `broadcast` is enabled
+	broadcastCh      chan types.TxKey
+	broadcastMtx     sync.Mutex
+	txsToBeBroadcast map[types.TxKey]struct{}
+}
+
+// NewTxPool constructs a new, empty content addressable txpool at the specified
+// initial height and using the given config and options.
+func NewTxPool(
+	logger log.Logger,
+	cfg *config.MempoolConfig,
+	proxyAppConn proxy.AppConnMempool,
+	height int64,
+	options ...TxPoolOption,
+) *TxPool {
+	txmp := &TxPool{
+		logger:           logger,
+		config:           cfg,
+		proxyAppConn:     proxyAppConn,
+		metrics:          mempool.NopMetrics(),
+		rejectedTxCache:  NewLRUTxCache(cfg.CacheSize),
+		evictedTxs:       NewEvictedTxCache(evictedTxCacheSize),
+		seenByPeersSet:   NewSeenTxSet(),
+		height:           height,
+		preCheckFn:       func(_ types.Tx) error { return nil },
+		postCheckFn:      func(_ types.Tx, _ *abci.ResponseCheckTx) error { return nil },
+		store:            newStore(),
+		broadcastCh:      make(chan types.TxKey, 1),
```

Comment:
Considering that `broadcastCh` is unbuffered (based on the description of the `broadcastCh` field above), we shall skip the capacity i.e.,
```suggestion
		broadcastCh:      make(chan types.TxKey),
```

Reply:
Yeah that makes sense